### PR TITLE
util: use go-deadlock to find deadlock

### LINKED ISSUE #40293: add go-deadlock to find the deadlock in the unit test
## Enhancement

In https://github.com/pingcap/tidb/pull/39584, we invoke SetGlobal of variable A in the Validation of variable B. This brought a circular dependency in rebuildSysVarCache. It will lead to deadlock. so we need a way to find the deadlock in the CI as soon as possible.

### PR DESCRIPTION
**Pull request by hawkingrei**
Signed-off-by: Weizhen Wang <wangweizhen@pingcap.com>

### What problem does this PR solve?

Issue Number: close #40293

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

We can find this log in the CI. It has been fixed in the https://github.com/pingcap/tidb/issues/40240
```
POTENTIAL DEADLOCK:

Previous place where the lock was grabbed

goroutine 273418 lock 0xc005e863b0

domain/sysvar_cache.go:121 domain.(*Domain).rebuildSysVarCache ??? <<<<<

domain/sysvar_cache.go:51 domain.(*Domain).rebuildSysVarCacheIfNeeded ???

domain/sysvar_cache.go:75 domain.(*Domain).GetGlobalVar ???

domain/domain.go:1541 domain.(*Domain).globalBindHandleWorkerLoop.func1 ???


Have been trying to lock it again for more than 20s

goroutine 261135 lock 0xc005e863b0

domain/sysvar_cache.go:121 domain.(*Domain).rebuildSysVarCache ??? <<<<<

domain/domain.go:2199 domain.(*Domain).NotifyUpdateSysVarCache ???

session/session.go:1450 session.(*session).replaceGlobalVariablesTableValue ???

session/session.go:1528 session.(*session).SetGlobalSysVarOnly ???

sessionctx/variable/varsutil.go:536 variable.updatePasswordValidationLength ???

sessionctx/variable/sysvar.go:553 variable.glob..func109 ???

sessionctx/variable/variable.go:361 variable.(*SysVar).ValidateWithRelaxedValidation ???

domain/sysvar_cache.go:146 domain.(*Domain).rebuildSysVarCache ???

domain/sysvar_cache.go:51 domain.(*Domain).rebuildSysVarCacheIfNeeded ???

domain/sysvar_cache.go:62 domain.(*Domain).GetSessionCache ???

session/session.go:3651 session.(*session).loadCommonGlobalVariablesIfNeeded ???

session/session.go:2140 session.(*session).ExecuteStmt ???

session/session.go:1674 session.(*session).ExecuteInternal ???

domain/domain.go:1310 domain.(*Domain).LoadPrivilegeLoop ???

session/session.go:3329 session.BootstrapSession ???

testkit/mockstore.go:84 testkit.bootstrap ???

testkit/mockstore.go:69 testkit.CreateMockStoreAndDomain ???

testkit/mockstore.go:61 testkit.CreateMockStore ???

executor/union_scan_test.go:419 executor_test.TestIssue28073 ???
```
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by lance6716 — APPROVED** [review]

**Comment by hawkingrei** [comment]
@bb7133 PTAL

File: util/syncutil/mutex_deadlock.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// Copyright 2022 PingCAP, Inc.
```
Comment (bb7133) [inline util/syncutil/mutex_deadlock.go:1]:
```suggestion
// Copyright 2023 PingCAP, Inc.
```
Ditto for the others.

File: session/session.go [hunk 2]
```diff
@@ -274,7 +275,7 @@ type session struct {
 	idxUsageCollector *handle.SessionIndexUsageCollector
 
 	functionUsageMu struct {
-		sync.RWMutex
+		syncutil.RWMutex
```
Comment (bb7133) [inline session/session.go:278]: Do you have a plan for the rest of all other `sync.RWMutex` usage in TiDB?

File: util/syncutil/mutex_deadlock.go (hunk 1 above)
Comment (hawkingrei) [reply util/syncutil/mutex_deadlock.go]: fix for all new files.

**Review by bb7133 — APPROVED** [review]
LGTM

File: session/session.go (hunk 2 above)
Comment (hawkingrei) [reply session/session.go:278]: Yes, But deadlock detection will slow the test so I need to slowly transition to ```syncutil```.

**Pull request merged**